doc/introduction/measurements.rst

```diff
@@ -524,6 +524,85 @@ Collecting statistics for sequences of mid-circuit measurements is supported wit
+      :func:`~pennylane.dynamic_one_shot` transforms must not be applied directly to a :class:`~pennylane.QNode`
+      as it can lead to incorrect behaviour.
+
+* ``postselect_mode``: To configure how invalid shots are handled when postselecting mid-circuit measurements
+  with finite-shot circuits. Use ``postselect_mode="hw-like"`` to discard invalid samples. In this case, the number
+  of samples that are used for processing results will be less than or equal to the total number of shots. If
+  ``postselect_mode="fill-shots"`` is used, then the postselected value will be picked unconditionally, and all
+  samples will be considered valid. The default behaviour is ``postselect_mode="hw-like"``.
+
+  .. code-block:: python3
+
+      import pennylane as qml
+      import numpy as np
+
+      dev = qml.device("default.qubit", wires=3, shots=10)
+
+      def circuit(x):
+          qml.RX(x, 0)
+          m0 = qml.measure(0, postselect=1)
+          qml.CNOT([0, 1])
+          return qml.sample(qml.PauliZ(0))
+
+      fill_shots_qnode = qml.QNode(circuit, dev, mcm_method="one-shot", postselect_mode="fill-shots")
+      hw_like_qnode = qml.QNode(circuit, dev, mcm_method="one-shot", postselect_mode="hw-like")
+
+  >>> fill_shots_qnode(np.pi / 2)
+  array([-1., -1., -1., -1., -1., -1., -1., -1., -1., -1.])
+  >>> hw_like_qnode(np.pi / 2)
+  array([-1., -1., -1., -1., -1., -1., -1.])
+
+  .. note::
+
+      When using the ``jax`` interface, ``postselect_mode="hw-like"`` will have different behaviour based on the
+      chosen ``mcm_method``.
+
+      * If ``mcm_method="one-shot"``, invalid shots will not be discarded. Instead, invalid samples will be replaced
+        by ``np.iinfo(np.int32).min``. These invalid samples will not be used for processing final results (like
+        expectation values), but will appear in the ``QNode`` output if samples are requested directly. Consider
+        the circuit below:
+
+        .. code-block:: python3
+
+            import pennylane as qml
+            import jax
+            import jax.numpy as jnp
+
+            dev = qml.device("default.qubit", wires=3, shots=10, seed=jax.random.PRNGKey(123))
+
+            @qml.qnode(dev, postselect_mode="hw-like", mcm_method="one-shot")
+            def circuit(x):
+                qml.RX(x, 0)
+                qml.measure(0, postselect=1)
+                return qml.sample(qml.PauliZ(0))
+
+        >>> x = jnp.array(1.8)
+        >>> f(x)
+        Array([-2.1474836e+09, -1.0000000e+00, -2.1474836e+09, -2.1474836e+09,
+               -1.0000000e+00, -2.1474836e+09, -1.0000000e+00, -2.1474836e+09,
+               -1.0000000e+00, -1.0000000e+00], dtype=float32, weak_type=True)
+
+      * If ``mcm_method="deferred"``, then using ``postselect_mode="hw-like"`` will have the same behaviour as when
+        ``postselect_mode="fill-shots"``. This is due to the limitations of the :func:`~pennylane.defer_measurements`
```

Comment:
I'm confused by this. I ran all combinations of using JAX or not, both `postselect_mode`s and both `mcm_method`s and it seems that JAX + `"hw-like"` + `"deferred"` is the same as (not JAX) + `"hw-like"` + `"deferred"`, but differs from JAX + `"fill-shots"` + `"deferred"`.
As a consequence, JAX + `"hw-like"` + `"deferred"` does not seem to be JIT compatible (because it actually returns a dynamic shape.

Reply:
Chatted offline. Posting response here for visibility.

If you use diff_method=None, you will see the expected behaviour with JAX-JIT. diff_method defaults to parameter shift with sampling, in which case tape parameters and converted to numpy and the execution happens inside a pure_callback. I don't really have a way of checking if we're jitting in that context. If you have any suggestions, that would be really helpful!

Reply:
This behaviour can definitely use better documentation though, I will add that. Also realized we can smartly change the `postselect_mode` to `"fill-shots"` in `qml.execute` if the interface of the tape is `"jax-jit"` and `postselect_mode=="hw-like"` and `diff_method != None`.

Reply:
I've updated `qml.execute` to change `postselect_mode` to `fill-shots` if jitting regardless of `diff_method`.

Reply:
Good call to remove that manual dependency on `diff_method` :sweat_smile: